R/data.table.R

```diff
@@ -1336,7 +1336,11 @@ replace_dot_alias = function(e) {
 
     if (is.data.table(jval)) {
       setattr(jval, 'class', class(x)) # fix for #64
-      if (haskey(x) && all(key(x) %chin% names(jval)) && suppressWarnings(is.sorted(jval, by=key(x))))  # TO DO: perhaps this usage of is.sorted should be allowed internally then (tidy up and make efficient)
+      if (haskey(x) && 
+          all(key(x) %chin% names(jval)) &&
+          ((is.null(irows) && jsub %iscall% "list" && all(vapply_1b(as.list(jsub)[-1L], is.name))) || ##fix for #4498
```

Comment:
this misses something like key=V1 and j=.(V1, f(V2)), as well as like j=c(list(V1=V1), .SD).

the latter is messy, maybe best to let is.sorted happen there.

I was thinking to add this check much earlier, around when j is first analyzed, there must be some similar loop done over j where we can tack this on.

FWIW both logics miss something like key=V1, j=.(newV = V1), but again, not sure how worth pursuing

Reply:
Agree, we have much more reliable way to figure that out. Better to be more conservative and not retain key when, we cannot be sure. Instead of recalculating order.